File: vite-project-react18/src/figma/figma-json-to-react/scripts/inspect_figma_rest_json.py
#!/usr/bin/env python3
"""Validate and summarize a local Figma REST API JSON file for React translation."""

from __future__ import annotations

import argparse
import json
from collections import Counter
from pathlib import Path
from typing import Any, NoReturn

VECTOR_LIKE_TYPES = {
    "VECTOR",
    "BOOLEAN_OPERATION",
    "STAR",
    "LINE",
    "ELLIPSE",
    "POLYGON",
}


def parse_data_url_mime_type(value: str) -> str | None:
    if not value.startswith("data:"):
        return None

    header = value.split(",", 1)[0]
    mime_type = header[5:].split(";", 1)[0]
    return mime_type or None


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Validate and summarize a local Figma REST API JSON file."
    )
    parser.add_argument("json_path", help="Absolute path to the input JSON file.")
    parser.add_argument(
        "--node-id",
        help="Optional node id when the JSON contains multiple candidate documents.",
    )
    parser.add_argument(
        "--max-depth",
        type=int,
        default=None,
        help="Optional maximum child depth for the summarized tree.",
    )
    parser.add_argument(
        "--output",
        help="Absolute path for the summarized JSON output.",
    )
    return parser.parse_args()


def fail(message: str) -> NoReturn:
    raise SystemExit(message)


def ensure_absolute_path(path_str: str, label: str) -> Path:
    path = Path(path_str)
    if not path.is_absolute():
        fail(f"{label} must be an absolute path: {path_str}")
    return path


def load_json(json_path: Path) -> dict[str, Any]:
    if not json_path.exists():
        fail(f"Input JSON file does not exist: {json_path}")
    if not json_path.is_file():
        fail(f"Input JSON path is not a file: {json_path}")

    try:
        payload = json.loads(json_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        fail(f"Input JSON file is not valid JSON: {exc}")

    if not isinstance(payload, dict):
        fail("Unsupported Figma JSON shape: top-level value must be an object.")
    return payload


def extract_node_entry(node_entry: Any, node_id: str) -> dict[str, Any]:
    if not isinstance(node_entry, dict):
        fail(f"Node entry for {node_id} is not an object.")
    document = node_entry.get("document")
    if not isinstance(document, dict):
        fail(f"Node entry for {node_id} does not contain a document object.")
    return document


def resolve_document(
    payload: dict[str, Any], requested_node_id: str | None
) -> tuple[str | None, str, dict[str, Any]]:
    wrapped_payload = payload.get("payload")
    if isinstance(wrapped_payload, dict):
        return resolve_document(wrapped_payload, requested_node_id)

    nodes = payload.get("nodes")
    if isinstance(nodes, dict):
        if requested_node_id:
            if requested_node_id not in nodes:
                fail(f"Requested node id not found in JSON: {requested_node_id}")
            return (
                requested_node_id,
                "nodes",
                extract_node_entry(nodes[requested_node_id], requested_node_id),
            )

        candidates = []
        for node_id, node_entry in nodes.items():
            if isinstance(node_entry, dict) and isinstance(node_entry.get("document"), dict):
                candidates.append((node_id, node_entry["document"]))

        if not candidates:
            fail("The JSON contains a nodes object but no document entries.")
        if len(candidates) > 1:
            available = ", ".join(node_id for node_id, _ in candidates[:20])
            fail(
                "Multiple node documents were found. "
                f"Pass --node-id to select one. Available node ids: {available}"
            )

        node_id, document = candidates[0]
        return node_id, "nodes", document

    document = payload.get("document")
    if isinstance(document, dict):
        selected_node_id = requested_node_id or payload.get("node_id") or document.get("id")
        return selected_node_id, "document", document

    if isinstance(payload.get("name"), str) and isinstance(payload.get("type"), str):
        selected_node_id = requested_node_id or payload.get("id")
        return selected_node_id, "direct-node", payload

    fail(
        "Unsupported Figma JSON shape. Expected a top-level document, "
        "a payload.document, a nodes map, or a simplified top-level node."
    )


def simplify_paint(paint: dict[str, Any]) -> dict[str, Any]:
    keys = [
        "type",
        "visible",
        "opacity",
        "blendMode",
        "scaleMode",
        "imageRef",
        "color",
        "gradientStops",
    ]
    return {key: paint[key] for key in keys if key in paint}


def simplify_effect(effect: dict[str, Any]) -> dict[str, Any]:
    keys = [
        "type",
        "visible",
        "radius",
        "spread",
        "blendMode",
        "offset",
        "color",
    ]
    return {key: effect[key] for key in keys if key in effect}


def simplify_style(style: dict[str, Any]) -> dict[str, Any]:
    keys = [
        "fontFamily",
        "fontPostScriptName",
        "fontWeight",
        "fontSize",
        "textAlignHorizontal",
        "textAlignVertical",
        "letterSpacing",
        "lineHeightPx",
        "lineHeightPercent",
        "lineHeightPercentFontSize",
        "lineHeightUnit",
        "textCase",
        "textDecoration",
    ]
    return {key: style[key] for key in keys if key in style}


def simplify_node(
    node: dict[str, Any], depth: int = 0, max_depth: int | None = None
) -> dict[str, Any]:
    keys = [
        "id",
        "name",
        "type",
        "visible",
        "opacity",
        "blendMode",
        "clipsContent",
        "layoutMode",
        "primaryAxisSizingMode",
        "counterAxisSizingMode",
        "primaryAxisAlignItems",
        "counterAxisAlignItems",
        "itemSpacing",
        "layoutAlign",
        "layoutGrow",
        "layoutWrap",
        "paddingLeft",
        "paddingRight",
        "paddingTop",
        "paddingBottom",
        "cornerRadius",
        "topLeftRadius",
        "topRightRadius",
        "bottomLeftRadius",
        "bottomRightRadius",
        "strokeWeight",
        "strokeAlign",
        "strokeJoin",
        "strokeCap",
        "characters",
        "constraints",
        "componentPropertyDefinitions",
        "componentProperties",
        "overflowDirection",
    ]

    cleaned = {key: node[key] for key in keys if key in node}

    if "absoluteBoundingBox" in node:
        cleaned["absoluteBoundingBox"] = node["absoluteBoundingBox"]
    if "absoluteRenderBounds" in node:
        cleaned["absoluteRenderBounds"] = node["absoluteRenderBounds"]
    if "fills" in node:
        cleaned["fills"] = [
            simplify_paint(paint) for paint in node["fills"] if isinstance(paint, dict)
        ]
    if "strokes" in node:
        cleaned["strokes"] = [
            simplify_paint(paint) for paint in node["strokes"] if isinstance(paint, dict)
        ]
    if "effects" in node:
        cleaned["effects"] = [
            simplify_effect(effect)
            for effect in node["effects"]
            if isinstance(effect, dict)
        ]
    if "style" in node and isinstance(node["style"], dict):
        cleaned["style"] = simplify_style(node["style"])
    image_url = node.get("imageUrl")
    if isinstance(image_url, str) and image_url:
        cleaned["hasImageUrl"] = True
        mime_type = parse_data_url_mime_type(image_url)
        if mime_type:
            cleaned["imageUrlMimeType"] = mime_type
    svg_content = node.get("svgContent")
    if isinstance(svg_content, str) and svg_content.strip():
        cleaned["hasSvgContent"] = True

    children = node.get("children")
    if isinstance(children, list):
        cleaned["childCount"] = len(children)
        if max_depth is None or depth < max_depth:
            cleaned["children"] = [
                simplify_node(child, depth=depth + 1, max_depth=max_depth)
                for child in children
                if isinstance(child, dict)
            ]

    return cleaned


def walk_nodes(node: dict[str, Any]) -> list[dict[str, Any]]:
    nodes = [node]
    children = node.get("children")
    if isinstance(children, list):
        for child in children:
            if isinstance(child, dict):
                nodes.extend(walk_nodes(child))
    return nodes


def collect_stats(document: dict[str, Any]) -> dict[str, Any]:
    nodes = walk_nodes(document)
    node_types = Counter()
    image_refs: set[str] = set()
    text_nodes = 0
    vector_nodes = 0
    embedded_image_nodes: list[dict[str, Any]] = []
    embedded_svg_nodes: list[dict[str, Any]] = []

    for node in nodes:
        node_type = node.get("type")
        if isinstance(node_type, str):
            node_types[node_type] += 1
            if node_type == "TEXT":
                text_nodes += 1
            if node_type in VECTOR_LIKE_TYPES:
                vector_nodes += 1

        fills = node.get("fills")
        if isinstance(fills, list):
            for fill in fills:
                if not isinstance(fill, dict):
                    continue
                image_ref = fill.get("imageRef")
                if isinstance(image_ref, str) and image_ref:
                    image_refs.add(image_ref)

        image_url = node.get("imageUrl")
        if isinstance(image_url, str) and image_url:
            embedded_image_nodes.append(
                {
                    "id": node.get("id"),
                    "name": node.get("name"),
                    "mime_type": parse_data_url_mime_type(image_url),
                }
            )

        svg_content = node.get("svgContent")
        if isinstance(svg_content, str) and svg_content.strip():
            embedded_svg_nodes.append(
                {
                    "id": node.get("id"),
                    "name": node.get("name"),
                }
            )

    return {
        "total_nodes": len(nodes),
        "text_nodes": text_nodes,
        "vector_like_nodes": vector_nodes,
        "image_fill_refs": sorted(image_refs),
        "embedded_image_url_count": len(embedded_image_nodes),
        "embedded_image_url_nodes": embedded_image_nodes[:20],
        "embedded_svg_count": len(embedded_svg_nodes),
        "embedded_svg_nodes": embedded_svg_nodes[:20],
        "node_types": dict(sorted(node_types.items())),
    }


def write_json(payload: dict[str, Any], output_path: Path | None) -> None:
    content = json.dumps(payload, ensure_ascii=False, indent=2)
    if output_path is None:
        print(content)
        return

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(content, encoding="utf-8")
    print(f"Wrote JSON summary to {output_path}")


def main() -> None:
    args = parse_args()
    json_path = ensure_absolute_path(args.json_path, "Input JSON path")
    output_path = ensure_absolute_path(args.output, "Output path") if args.output else None

    payload = load_json(json_path)
    node_id, source_format, document = resolve_document(payload, args.node_id)
    summary = {
        "input_path": str(json_path),
        "selected_node_id": node_id,
        "source_format": source_format,
        "stats": collect_stats(document),
        "document": simplify_node(document, max_depth=args.max_depth),
    }
    write_json(summary, output_path)


if __name__ == "__main__":
    main()

File: .codex/skills/leon-figma/scripts/inspect_figma_rest_json.js
#!/usr/bin/env node

const fs = require('node:fs');
const path = require('node:path');

const VECTOR_LIKE_TYPES = new Set([
  'VECTOR',
  'BOOLEAN_OPERATION',
  'STAR',
  'LINE',
  'ELLIPSE',
  'POLYGON',
]);
const DIRECT_SVG_NODE_TYPES = new Set(['VECTOR', 'BOOLEAN_OPERATION', 'STAR', 'POLYGON']);
const BASIC_VECTOR_NODE_TYPES = new Set(['ELLIPSE', 'LINE']);
const SVG_CONTAINER_TYPES = new Set(['GROUP', 'FRAME', 'COMPONENT', 'INSTANCE']);
const SVG_ELIGIBLE_NODE_TYPES = new Set([
  ...DIRECT_SVG_NODE_TYPES,
  ...BASIC_VECTOR_NODE_TYPES,
  ...SVG_CONTAINER_TYPES,
]);

function fail(message) {
  console.error(message);
  process.exit(1);
}

function parseDataUrlMimeType(value) {
  if (!value.startsWith('data:')) {
    return null;
  }
  const header = value.split(',', 1)[0];
  const mimeType = header.slice(5).split(';', 1)[0];
  return mimeType || null;
}

function parseArgs(argv) {
  const args = {
    jsonPath: undefined,
    nodeId: undefined,
    maxDepth: undefined,
    output: undefined,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === '--node-id') {
      index += 1;
      args.nodeId = readOptionValue(argv, index, arg);
      continue;
    }
    if (arg === '--max-depth') {
      index += 1;
      args.maxDepth = Number.parseInt(readOptionValue(argv, index, arg), 10);
      continue;
    }
    if (arg === '--output') {
      index += 1;
      args.output = readOptionValue(argv, index, arg);
      continue;
    }
    if (arg.startsWith('--')) {
      fail(`Unknown argument: ${arg}`);
    }
    if (args.jsonPath) {
      fail(`Unexpected positional argument: ${arg}`);
    }
    args.jsonPath = arg;
  }

  if (!args.jsonPath) {
    fail('Missing JSON path.');
  }
  if (args.maxDepth !== undefined && !Number.isInteger(args.maxDepth)) {
    fail('--max-depth must be an integer.');
  }
  return args;
}

function readOptionValue(argv, index, optionName) {
  if (index >= argv.length) {
    fail(`Missing value for ${optionName}`);
  }
  return argv[index];
}

function isAbsolutePath(value) {
  return path.isAbsolute(value) || path.win32.isAbsolute(value);
}

function ensureAbsolutePath(value, label) {
  if (!isAbsolutePath(value)) {
    fail(`${label} must be an absolute path: ${value}`);
  }
  return value;
}

function loadJson(jsonPath) {
  if (!fs.existsSync(jsonPath)) {
    fail(`Input JSON file does not exist: ${jsonPath}`);
  }
  if (!fs.statSync(jsonPath).isFile()) {
    fail(`Input JSON path is not a file: ${jsonPath}`);
  }
  try {
    const payload = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
    if (!isPlainObject(payload)) {
      fail('Unsupported Figma JSON shape: top-level value must be an object.');
    }
    return payload;
  } catch (error) {
    fail(`Input JSON file is not valid JSON: ${error.message}`);
  }
}

function isPlainObject(value) {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}

function extractNodeEntry(nodeEntry, nodeId) {
  if (!isPlainObject(nodeEntry)) {
    fail(`Node entry for ${nodeId} is not an object.`);
  }
  if (!isPlainObject(nodeEntry.document)) {
    fail(`Node entry for ${nodeId} does not contain a document object.`);
  }
  return nodeEntry.document;
}

function resolveDocument(payload, requestedNodeId) {
  if (isPlainObject(payload.payload)) {
    return resolveDocument(payload.payload, requestedNodeId);
  }
  if (isPlainObject(payload.nodes)) {
    return resolveNodesDocument(payload.nodes, requestedNodeId);
  }
  if (isPlainObject(payload.document)) {
    return {
      nodeId: requestedNodeId || payload.node_id || payload.document.id,
      sourceFormat: 'document',
      document: payload.document,
    };
  }
  if (typeof payload.name === 'string' && typeof payload.type === 'string') {
    return {
      nodeId: requestedNodeId || payload.id,
      sourceFormat: 'direct-node',
      document: payload,
    };
  }
  fail('Unsupported Figma JSON shape. Expected a top-level document, a payload.document, a nodes map, or a simplified top-level node.');
}

function resolveNodesDocument(nodes, requestedNodeId) {
  if (requestedNodeId) {
    if (!Object.prototype.hasOwnProperty.call(nodes, requestedNodeId)) {
      fail(`Requested node id not found in JSON: ${requestedNodeId}`);
    }
    return {
      nodeId: requestedNodeId,
      sourceFormat: 'nodes',
      document: extractNodeEntry(nodes[requestedNodeId], requestedNodeId),
    };
  }

  const candidates = Object.entries(nodes)
    .filter(([, nodeEntry]) => isPlainObject(nodeEntry) && isPlainObject(nodeEntry.document))
    .map(([nodeId, nodeEntry]) => ({ nodeId, document: nodeEntry.document }));
  if (candidates.length === 0) {
    fail('The JSON contains a nodes object but no document entries.');
  }
  if (candidates.length > 1) {
    const available = candidates.slice(0, 20).map((candidate) => candidate.nodeId).join(', ');
    fail(`Multiple node documents were found. Pass --node-id to select one. Available node ids: ${available}`);
  }
  return {
    nodeId: candidates[0].nodeId,
    sourceFormat: 'nodes',
    document: candidates[0].document,
  };
}

function pickKeys(source, keys) {
  const picked = {};
  for (const key of keys) {
    if (Object.prototype.hasOwnProperty.call(source, key)) {
      picked[key] = source[key];
    }
  }
  return picked;
}

function simplifyPaint(paint) {
  return pickKeys(paint, [
    'type',
    'visible',
    'opacity',
    'blendMode',
    'scaleMode',
    'imageRef',
    'color',
    'gradientStops',
  ]);
}

function simplifyEffect(effect) {
  return pickKeys(effect, [
    'type',
    'visible',
    'radius',
    'spread',
    'blendMode',
    'offset',
    'color',
  ]);
}

function simplifyStyle(style) {
  return pickKeys(style, [
    'fontFamily',
    'fontPostScriptName',
    'fontWeight',
    'fontSize',
    'textAlignHorizontal',
    'textAlignVertical',
    'letterSpacing',
    'lineHeightPx',
    'lineHeightPercent',
    'lineHeightPercentFontSize',
    'lineHeightUnit',
    'textCase',
    'textDecoration',
  ]);
}

function simplifyNode(node, depth = 0, maxDepth = undefined) {
  const cleaned = pickKeys(node, [
    'id',
    'name',
    'type',
    'visible',
    'opacity',
    'blendMode',
    'clipsContent',
    'layoutMode',
    'primaryAxisSizingMode',
    'counterAxisSizingMode',
    'primaryAxisAlignItems',
    'counterAxisAlignItems',
    'itemSpacing',
    'layoutAlign',
    'layoutGrow',
    'layoutWrap',
    'paddingLeft',
    'paddingRight',
    'paddingTop',
    'paddingBottom',
    'cornerRadius',
    'topLeftRadius',
    'topRightRadius',
    'bottomLeftRadius',
    'bottomRightRadius',
    'strokeWeight',
    'strokeAlign',
    'strokeJoin',
    'strokeCap',
    'characters',
    'constraints',
    'componentPropertyDefinitions',
    'componentProperties',
    'overflowDirection',
  ]);
  addGeometry(node, cleaned);
  addVisualFields(node, cleaned);
  addEmbeddedAssetFlags(node, cleaned);
  addSimplifiedChildren(node, cleaned, depth, maxDepth);
  return cleaned;
}

function addGeometry(node, cleaned) {
  if (Object.prototype.hasOwnProperty.call(node, 'absoluteBoundingBox')) {
    cleaned.absoluteBoundingBox = node.absoluteBoundingBox;
  }
  if (Object.prototype.hasOwnProperty.call(node, 'absoluteRenderBounds')) {
    cleaned.absoluteRenderBounds = node.absoluteRenderBounds;
  }
}

function addVisualFields(node, cleaned) {
  if (Array.isArray(node.fills)) {
    cleaned.fills = node.fills.filter(isPlainObject).map(simplifyPaint);
  }
  if (Array.isArray(node.strokes)) {
    cleaned.strokes = node.strokes.filter(isPlainObject).map(simplifyPaint);
  }
  if (Array.isArray(node.effects)) {
    cleaned.effects = node.effects.filter(isPlainObject).map(simplifyEffect);
  }
  if (isPlainObject(node.style)) {
    cleaned.style = simplifyStyle(node.style);
  }
}

function addEmbeddedAssetFlags(node, cleaned) {
  if (typeof node.imageUrl === 'string' && node.imageUrl) {
    cleaned.hasImageUrl = true;
    const mimeType = parseDataUrlMimeType(node.imageUrl);
    if (mimeType) {
      cleaned.imageUrlMimeType = mimeType;
    }
  }
  if (typeof node.svgContent === 'string' && node.svgContent.trim()) {
    cleaned.hasSvgContent = true;
  }
}

function addSimplifiedChildren(node, cleaned, depth, maxDepth) {
  if (!Array.isArray(node.children)) {
    return;
  }
  cleaned.childCount = node.children.length;
  if (maxDepth === undefined || maxDepth === null || depth < maxDepth) {
    cleaned.children = node.children
      .filter(isPlainObject)
      .map((child) => simplifyNode(child, depth + 1, maxDepth));
  }
}

function walkNodes(node) {
  const nodes = [node];
  if (Array.isArray(node.children)) {
    for (const child of node.children) {
      if (isPlainObject(child)) {
        nodes.push(...walkNodes(child));
      }
    }
  }
  return nodes;
}

function iterPaints(node) {
  const paints = [];
  for (const paintKey of ['fills', 'strokes']) {
    if (Array.isArray(node[paintKey])) {
      paints.push(...node[paintKey].filter(isPlainObject));
    }
  }
  return paints;
}

function nodeHasRasterPaint(node) {
  return iterPaints(node).some((paint) => {
    if (paint.type === 'IMAGE') {
      return true;
    }
    return typeof paint.imageRef === 'string' && Boolean(paint.imageRef);
  });
}

function nodeHasTextContent(node) {
  if (node.type === 'TEXT') {
    return true;
  }
  return typeof node.characters === 'string' && Boolean(node.characters.trim());
}

function isExportableSvgCandidate(node, rootNodeId) {
  if (typeof node.id !== 'string' || node.id === rootNodeId) {
    return false;
  }
  if (typeof node.svgContent === 'string' && node.svgContent.trim()) {
    return true;
  }
  if (typeof node.type !== 'string' || !SVG_ELIGIBLE_NODE_TYPES.has(node.type)) {
    return false;
  }
  if (nodeHasTextContent(node) || nodeHasRasterPaint(node)) {
    return false;
  }
  const descendants = walkNodes(node).filter((descendant) => descendant !== node);
  if (DIRECT_SVG_NODE_TYPES.has(node.type)) {
    return true;
  }
  if (BASIC_VECTOR_NODE_TYPES.has(node.type) || descendants.length === 0) {
    return false;
  }
  return descendantTypesAllowSvgExport(descendants);
}

function descendantTypesAllowSvgExport(descendants) {
  const descendantTypes = new Set();
  for (const descendant of descendants) {
    if (!isValidSvgDescendant(descendant)) {
      return false;
    }
    descendantTypes.add(descendant.type);
  }
  return [...descendantTypes].some((type) => {
    return DIRECT_SVG_NODE_TYPES.has(type) || BASIC_VECTOR_NODE_TYPES.has(type);
  });
}

function isValidSvgDescendant(descendant) {
  if (typeof descendant.type !== 'string' || !SVG_ELIGIBLE_NODE_TYPES.has(descendant.type)) {
    return false;
  }
  if (nodeHasTextContent(descendant) || nodeHasRasterPaint(descendant)) {
    return false;
  }
  return !(typeof descendant.layoutMode === 'string' && descendant.layoutMode);
}

function collectSvgExportCandidates(document, rootNodeId) {
  const candidates = [];
  function visit(node) {
    if (isExportableSvgCandidate(node, rootNodeId)) {
      candidates.push({
        id: node.id,
        name: node.name,
        type: node.type,
        has_svg_content: typeof node.svgContent === 'string' && Boolean(node.svgContent.trim()),
        absolute_bounding_box: node.absoluteBoundingBox,
      });
      return;
    }
    if (Array.isArray(node.children)) {
      for (const child of node.children) {
        if (isPlainObject(child)) {
          visit(child);
        }
      }
    }
  }
  visit(document);
  return candidates;
}

function nodeUsesRasterAsset(node) {
  if (typeof node.imageUrl === 'string' && node.imageUrl) {
    return true;
  }
  return nodeHasRasterPaint(node);
}

function summarizeBoundaryNode(node) {
  const descendants = walkNodes(node).filter((descendant) => descendant !== node);
  return {
    id: node.id,
    name: node.name,
    type: node.type,
    layout_mode: node.layoutMode,
    child_count: Array.isArray(node.children) ? node.children.length : 0,
    absolute_bounding_box: node.absoluteBoundingBox,
    direct_raster_asset: nodeUsesRasterAsset(node),
    descendant_text_count: countDescendants(descendants, (descendant) => descendant.type === 'TEXT'),
    descendant_raster_count: countDescendants(descendants, nodeUsesRasterAsset),
    descendant_vector_count: countDescendants(descendants, isVectorLikeNode),
  };
}

function countDescendants(descendants, predicate) {
  return descendants.reduce((count, descendant) => {
    return predicate(descendant) ? count + 1 : count;
  }, 0);
}

function isVectorLikeNode(node) {
  return typeof node.type === 'string' && VECTOR_LIKE_TYPES.has(node.type);
}

function collectImplementationRegions(document) {
  if (!Array.isArray(document.children)) {
    return [];
  }
  const svgCandidateIds = new Set(
    collectSvgExportCandidates(document, document.id)
      .map((candidate) => candidate.id)
      .filter((id) => typeof id === 'string'),
  );
  return document.children
    .filter((child) => isImplementationRegion(child, svgCandidateIds))
    .map((child) => {
      const summary = summarizeBoundaryNode(child);
      summary.do_not_rasterize = true;
      summary.recommended_strategy = 'jsx-region';
      return summary;
    });
}

function isImplementationRegion(child, svgCandidateIds) {
  if (!isPlainObject(child)) {
    return false;
  }
  const childCount = Array.isArray(child.children) ? child.children.length : 0;
  if (childCount === 0 && nodeUsesRasterAsset(child)) {
    return false;
  }
  return !(childCount === 0 && typeof child.id === 'string' && svgCandidateIds.has(child.id));
}

function collectRasterAssetNodes(document) {
  return walkNodes(document)
    .filter(nodeUsesRasterAsset)
    .map((node) => {
      const summary = summarizeBoundaryNode(node);
      summary.recommended_strategy = 'leaf-image-asset';
      summary.is_leaf_asset = summary.child_count === 0;
      return summary;
    });
}

function collectTopLevelLayoutHints(document) {
  const root = readRootBounds(document);
  if (!root || !Array.isArray(document.children)) {
    return [];
  }
  return boundedChildren(document.children)
    .sort(compareNodeBounds)
    .reduce((state, child) => {
      const hint = buildLayoutHint(child, root, state.previousBottom);
      state.previousBottom = Math.max(state.previousBottom, hint.top + hint.height);
      state.hints.push(hint);
      return state;
    }, { previousBottom: 0, hints: [] }).hints;
}

function readRootBounds(document) {
  if (!isPlainObject(document.absoluteBoundingBox)) {
    return null;
  }
  const box = document.absoluteBoundingBox;
  const values = [box.x, box.y, box.width, box.height];
  if (!values.every((value) => typeof value === 'number')) {
    return null;
  }
  return {
    x: box.x,
    y: box.y,
    width: box.width,
    height: box.height,
  };
}

function boundedChildren(children) {
  return children.filter((child) => {
    return isPlainObject(child) && isPlainObject(child.absoluteBoundingBox);
  });
}

function compareNodeBounds(leftNode, rightNode) {
  const leftBox = leftNode.absoluteBoundingBox;
  const rightBox = rightNode.absoluteBoundingBox;
  return (leftBox.y || 0) - (rightBox.y || 0) || (leftBox.x || 0) - (rightBox.x || 0);
}

function buildLayoutHint(child, root, previousBottom) {
  const box = child.absoluteBoundingBox;
  const left = (box.x || 0) - root.x;
  const top = (box.y || 0) - root.y;
  const width = box.width || 0;
  const height = box.height || 0;
  const rightInset = root.width - (left + width);
  const bottomInset = root.height - (top + height);
  return {
    id: child.id,
    name: child.name,
    type: child.type,
    left,
    top,
    width,
    height,
    right_inset: rightInset,
    bottom_inset: bottomInset,
    gap_from_previous: top - previousBottom,
    horizontal_alignment: resolveHorizontalAlignment(left, rightInset, width, root.width),
  };
}

function resolveHorizontalAlignment(left, rightInset, width, rootWidth) {
  if (Math.abs(left) <= 1 && Math.abs(width - rootWidth) <= 1) {
    return 'full-width';
  }
  if (Math.abs(left - rightInset) <= 1) {
    return 'centered';
  }
  if (Math.abs(rightInset) <= 1) {
    return 'flush-right';
  }
  if (Math.abs(left) <= 1) {
    return 'flush-left';
  }
  return 'custom';
}

function collectStats(document) {
  const nodes = walkNodes(document);
  const imageRefs = new Set();
  const nodeTypes = new Map();
  const embeddedImageNodes = [];
  const embeddedSvgNodes = [];
  const counter = { textNodes: 0, vectorNodes: 0 };
  const svgExportCandidates = collectSvgExportCandidates(document, document.id);
  const implementationRegions = collectImplementationRegions(document);
  const rasterAssetNodes = collectRasterAssetNodes(document);
  const topLevelLayoutHints = collectTopLevelLayoutHints(document);

  for (const node of nodes) {
    collectNodeTypeStats(node, nodeTypes, counter);
    collectImageRefs(node, imageRefs);
    collectEmbeddedImageNode(node, embeddedImageNodes);
    collectEmbeddedSvgNode(node, embeddedSvgNodes);
  }

  return {
    total_nodes: nodes.length,
    text_nodes: counter.textNodes,
    vector_like_nodes: counter.vectorNodes,
    image_fill_refs: [...imageRefs].sort(),
    embedded_image_url_count: embeddedImageNodes.length,
    embedded_image_url_nodes: embeddedImageNodes.slice(0, 20),
    embedded_svg_count: embeddedSvgNodes.length,
    embedded_svg_nodes: embeddedSvgNodes.slice(0, 20),
    svg_export_candidate_count: svgExportCandidates.length,
    svg_export_candidates: svgExportCandidates.slice(0, 50),
    implementation_region_count: implementationRegions.length,
    implementation_regions: implementationRegions.slice(0, 50),
    raster_asset_node_count: rasterAssetNodes.length,
    raster_asset_nodes: rasterAssetNodes.slice(0, 50),
    non_leaf_raster_asset_node_count: rasterAssetNodes.filter((node) => !node.is_leaf_asset).length,
    top_level_layout_hints: topLevelLayoutHints.slice(0, 50),
    node_types: Object.fromEntries([...nodeTypes.entries()].sort(([left], [right]) => left.localeCompare(right))),
  };
}

function collectNodeTypeStats(node, nodeTypes, counter) {
  if (typeof node.type !== 'string') {
    return;
  }
  nodeTypes.set(node.type, (nodeTypes.get(node.type) || 0) + 1);
  if (node.type === 'TEXT') {
    counter.textNodes += 1;
  }
  if (VECTOR_LIKE_TYPES.has(node.type)) {
    counter.vectorNodes += 1;
  }
}

function collectImageRefs(node, imageRefs) {
  for (const paint of iterPaints(node)) {
    if (typeof paint.imageRef === 'string' && paint.imageRef) {
      imageRefs.add(paint.imageRef);
    }
  }
}

function collectEmbeddedImageNode(node, embeddedImageNodes) {
  if (typeof node.imageUrl !== 'string' || !node.imageUrl) {
    return;
  }
  embeddedImageNodes.push({
    id: node.id,
    name: node.name,
    mime_type: parseDataUrlMimeType(node.imageUrl),
  });
}

function collectEmbeddedSvgNode(node, embeddedSvgNodes) {
  if (typeof node.svgContent !== 'string' || !node.svgContent.trim()) {
    return;
  }
  embeddedSvgNodes.push({
    id: node.id,
    name: node.name,
  });
}

function writeJson(payload, outputPath) {
  const content = JSON.stringify(payload, null, 2);
  if (!outputPath) {
    console.log(content);
    return;
  }
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  fs.writeFileSync(outputPath, content, 'utf8');
  console.log(`Wrote JSON summary to ${outputPath}`);
}

function buildSummary(jsonPath, args, documentInfo) {
  const implementationRegions = collectImplementationRegions(documentInfo.document);
  const rasterAssetNodes = collectRasterAssetNodes(documentInfo.document);
  const svgExportCandidates = collectSvgExportCandidates(documentInfo.document, documentInfo.document.id);
  return {
    input_path: jsonPath,
    selected_node_id: documentInfo.nodeId,
    source_format: documentInfo.sourceFormat,
    stats: collectStats(documentInfo.document),
    layout_hints: {
      root_bounds: documentInfo.document.absoluteBoundingBox,
      top_level_nodes: collectTopLevelLayoutHints(documentInfo.document),
    },
    asset_boundaries: {
      implementation_regions: implementationRegions,
      do_not_rasterize_region_ids: idsFrom(implementationRegions),
      raster_asset_nodes: rasterAssetNodes,
      leaf_raster_asset_node_ids: leafRasterIdsFrom(rasterAssetNodes),
      svg_asset_nodes: svgExportCandidates,
      svg_asset_node_ids: idsFrom(svgExportCandidates),
    },
    document: simplifyNode(documentInfo.document, 0, args.maxDepth),
  };
}

function idsFrom(nodes) {
  return nodes.map((node) => node.id).filter((id) => typeof id === 'string');
}

function leafRasterIdsFrom(nodes) {
  return nodes
    .filter((node) => node.is_leaf_asset)
    .map((node) => node.id)
    .filter((id) => typeof id === 'string');
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const jsonPath = ensureAbsolutePath(args.jsonPath, 'Input JSON path');
  const outputPath = args.output ? ensureAbsolutePath(args.output, 'Output path') : undefined;
  const payload = loadJson(jsonPath);
  const documentInfo = resolveDocument(payload, args.nodeId);
  writeJson(buildSummary(jsonPath, args, documentInfo), outputPath);
}

main();
